### ISSUE TITLE
Adding `_max_workers` attribute to `MPIPoolExecutor` subclass

### ISSUE THREAD

**Issue by jakirkham** [mpi4py/mpi4py#94]
In `concurrent.futures`, the implementations define an attribute `_max_workers` for [the `ThreadPoolExecutor`]( https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/concurrent/futures/thread.py#L143 ) and [the `ProcessPoolExecutor`]( https://github.com/python/cpython/blob/d56e700d6c7ebf1a424260d0e4f0c291d1f5b3af/Lib/concurrent/futures/process.py#L596 ), which shows how many workers can be spun up with tasks at a time. It would be helpful if [`MPIPoolExecutor`]( https://github.com/mpi4py/mpi4py/blob/18f16da45eaaacfba4cc0293964e1a6548d4c6a6/src/mpi4py/futures/pool.py#L17 ) also defined `_max_workers` so this could be queried on them as well. This can help when deciding how many tasks to dispatch to the `Executor` or how to batch them.

**Comment by dalcinl** [comment]
I'm a bit hesitant to implement and support private APIs. Some internal details of `mpi4py.futures` are a bit different than `concurrent.futures`. For example, `concurrent.futures` create workers on-demand, adjusting the number of workers as tasks arrive and up to `max_workers`. `mpi4py.futures` asks the MPI runtime to spawn `max_workers` processes in "soft" mode, than means that the runtime could spawn less processes if the resources are not available. All this is MPI standard semantics (actual support and behavior depends on the backend MPI implementation). Therefore, a `_max_workers` attribute in `mpi4py.futures` could be misleading, that's the reason I do not expose it What you really want is the actual number of MPI worker processes, not the upper bound requested by users.

All that being said, I would like to offer you an alternative.

There is an easy way (albeit still private API) to know the number of MPI workers is the following:

```python
executor = MPIPoolExecutor(...)
executor.bootup()  # block until workers are up and running
num_workers = executor._pool.size  # this is the actual number of MPI workers
executor.submit(...)
```

We are using this approach for our own [testing](https://github.com/mpi4py/mpi4py/blob/master/demo/futures/test_futures.py#L325-L326).

If you want inject a `_max_workers` attribute for compatibility with `concurrent.futures`, then it is as easy picking one of the following options:

* Use a factory function to create the executor and set the `_max_workers` attribute in the executor instance.
* Subclass `MPIPoolExecutor` and add a `@property` getter to get  `_max_workers`.
* Monkey-patch the `MPIPoolExecutor` class by injecting a `property` getter to get `_max_workers`.

mpi4py could expose a convenience `num_workers` attribute (implemented via property getter). But then we need to decide about the semantics (return value) in between the executor is created and the workers and up and running (return None? block until workers are up and return the number of workers?). FYI, creating an executor does not block until the workers have been spawned (that's the reason for the `bootup()` call in the code above). Also, we need to decide on the semantics after the executor has shutdown (return None? return the number of workers the executor used while alive? error with exception?).

**Comment by dalcinl** [comment]
@jakirkham Any thoughts? Anything to add? Is the workaround enough for you?

**Comment by jakirkham** [comment]
Thanks for the context, Lisandro 🙂 Sorry for the slow reply

Not quite. Perhaps it helps to see my use case?

In Dask we added support somewhat recently for [using `concurrent.futures`-based `Executor`s]( https://docs.dask.org/en/latest/scheduling.html#configuration ). By default this will use the `ThreadPoolExecutor`, but users could switch to processes, which would use the `ProcessPoolExecutor`. This becomes more interesting when users bring their own `Executor` to support other environments ( https://github.com/dask/dask/issues/6220 ). As a result we can help users get started with Dask in environments they know well as long as there is an `Executor` they can leverage.

MPI support is a commonly requested thing. We do have [Dask-MPI]( https://mpi.dask.org/en/latest/ ), which does support launching the Distributed Dask cluster with `mpi4py`. This may be a bit heavyweight for a user, who is just trying to get started and play around a little bit. Being able to start a `MPIPoolExecutor` reasonably quickly and plug that into Dask, could be helpful here.

However when scheduling work, it can be useful to know [how many workers]( https://docs.dask.org/en/latest/scheduler-overview.html?highlight=num_workers#configuring-the-schedulers ) can be used. This helps us make decisions about how to batch work to an individual worker or how many tasks we can reasonably run. Unfortunately we can't really get by without this information. For most `Executor`s, we have seen `_max_workers` works well or can be easily added. So [we check for that]( https://github.com/dask/dask/blob/618aedf8ddc4db71ca19db870421d3be646d9da1/dask/base.py#L1178-L1181 ). However we are left to guess without that information.

So this leaves us with this question, how should we try to support `MPIPoolExecutor` here? From Dask's perspective `_max_workers` would be ideal as this would impose the least effort on users. If `executor._pool.size` were aliased that way, that would certainly do the trick. Though it sounds like you have reservations with this approach. Within Dask, we could add more specific checks with different `Executor`s, but this likely won't scale well and we may not be in the best position to maintain these changes as implementations evolve. If neither of those work, we could document this way of getting the number of MPI workers or simply point users to this thread. Then users would run something like the code below (after the startup code you have above):

```python
with dask.config.set(scheduler=executor, num_workers=executor._pool.size):
    x.compute()
```

Maybe this is an ok place to leave things? This is admittedly a relatively new use case. So it may simply need more users trying things and providing feedback about how they would like things to work. That would hopefully clarify any next steps here that we are unsure of atm

**Comment by leofang** [comment]
Interesting. @dalcinl I just read the 4.0 spec and noticed the following description (Sec. 11.8.2):
> If you want the behavior “spawn as many processes as possible, up to N,” you should do a soft spawn, where the set of allowed values {mi} is {0, . . . , N}. However, this is not completely portable, as implementations are not required to support soft spawning.

If soft-spawn is not portable, perhaps we should not set it as default? By doing so, we get `_max_workers` working as intended for free IIUC.

**Comment by dalcinl** [comment]
I'm hesitant to trade a IMHO useful, *public* and documented, albeit optional, feature of MPI in favor of a  *private* Python API. If MPI implementation do not support soft spawn if perhaps just out of laziness/oversight, I don't see any fundamental complications on supporting the feature.

The `_max_worker` attribute can be supported with a few caveats and decisions.

1. Getting `_max_workers` should block the calling thread until the MPI worker processes are up and running. You have to pay to get it, no free luch.
2. What to do if `initializer` fails? Return `None`? Return `0`? Raise the `BrokenExecutor` exception?
3. What should be the behavior of getting the attribute after shutdown? Return `None`? Return the value it had before shutdown? Raise exception?

**Comment by jakirkham** [comment]
Closing now that PR ( https://github.com/mpi4py/mpi4py/pull/117 ) adds this

**Comment by dalcinl** [comment]
@jakirkham Please note that this feature will be available in the next patch release 3.1.2. I'm waiting for Python 3.10 to propagate to the many CI services (particularly AppVeyor, where I build Windows wheels).
